### ISSUE TITLE
SUGGESTION: Fool-proof scheme.

### ISSUE THREAD

**Issue by Misha4ek** [DiffuseHyperion/InertiaAntiCheat#48]
**Is your feature request related to a problem? Please describe.**
A friend of mine manage to bypass Inertia anti-cheat by modifying name in mod files so it mimics some client side mod.

**Describe the solution you'd like**
I want for Inertia anti-cheat to check code in mod files, not the size. Size can be artificially increased by for example 8 billions letters "a" as comment in the code.

**Additional context**
I may create how-to-do video by request.

**Comment by DiffuseHyperion** [comment]
The problem you described should only be an issue in the alpha versions. On more modern versions, the mod no longer only checks for the name but instead the actual file, so you can't just rename the file.

**Comment by Misha4ek** [comment]
Well, I host the server on 1.20.4, and I guess I just wait for the backport?

**Comment by Tunga-Tech** [comment]
It should just zip up the full mod list and upload it to the server.

**Comment by huzpsb** [comment]
> The problem you described should only be an issue in the alpha versions. On more modern versions, the mod no longer only checks for the name but instead the actual file, so you can't just rename the file.

But it's still easily bypassable... I mean, let's insert a "0" here
https://github.com/DiffuseHyperion/InertiaAntiCheat/blob/376e2ae3234347b05658bcffdbf79b48edf08bfb/src/main/java/me/diffusehyperion/inertiaanticheat/client/InertiaAntiCheatClient.java#L31

It's then
````java
File modDirectory = FabricLoader.getInstance().getGameDir().resolve("mods1").toFile();
````

So let's just
mv mod mod1
And do whatever you want

**Comment by Misha4ek** [comment]
> > The problem you described should only be an issue in the alpha versions. On more modern versions, the mod no longer only checks for the name but instead the actual file, so you can't just rename the file.
> 
> But it's still easily bypassable... I mean, let's insert a "1" here
> 
> https://github.com/DiffuseHyperion/InertiaAntiCheat/blob/376e2ae3234347b05658bcffdbf79b48edf08bfb/src/main/java/me/diffusehyperion/inertiaanticheat/client/InertiaAntiCheatClient.java#L31
> 
> It's then
> 
> ```java
> File modDirectory = FabricLoader.getInstance().getGameDir().resolve("mods1").toFile();
> ```
> 
> So let's just mv mod mod1 And do whatever you want

We need for inertia anticheat to check itself's hash, and compare it with server one to fix this problem. I am right?

**Comment by DiffuseHyperion** [comment]
Both of you are correct here. It is possible to bypass the anticheat by recompiling the mod with the change mentioned above, and it would help if Inertia checked for its own hash, which is actually quite possible to do since Modrinth's API allows you to find the hash of released files ([example for version 1.0.4](https://api.modrinth.com/v2/version/gdSvdQ3F)). 

Ultimately I don't think the mod can be entirely foolproof as the server has no way to verify the legitimacy of the client's mod. There is no existing code in Fabric API that allows the server to get the client's modlist, so the only option is to trust the client in whatever it says, which isn't 100% safe since the mod could be recompiled with changes as mentioned earlier. In fact, the fix stated earlier could be bypassed by just having the malicious mod report a hardcoded hash to the server.

As stated in the mod description, the mod only really works on people who just download cheat mods without actual coding knowledge, which hopefully is most people.

**Comment by huzpsb** [comment]
> Both of you are correct here. It is possible to bypass the anticheat by recompiling the mod with the change mentioned above, and it would help if Inertia checked for its own hash, which is actually quite possible to do since Modrinth's API allows you to find the hash of released files ([example for version 1.0.4](https://api.modrinth.com/v2/version/gdSvdQ3F)). 
> 
> Ultimately I don't think the mod can be entirely foolproof as the server has no way to verify the legitimacy of the client's mod. There is no existing code in Fabric API that allows the server to get the client's modlist, so the only option is to trust the client in whatever it says, which isn't 100% safe since the mod could be recompiled with changes as mentioned earlier. In fact, the fix stated earlier could be bypassed by just having the malicious mod report a hardcoded hash to the server.
> 
> As stated in the mod description, the mod only really works on people who just download cheat mods without actual coding knowledge, which hopefully is most people.

Sorry for reopening this issue but I felt obliged to do so.
There is no way to make a anti-cheat mod fully fool-proof, but there are ways to make it "harder".
I'll list some commonly seen ways:
- Simply make it close sourced and add heavy obfuscations (which I'm actually against but quite effective)
- Add some checks that required actual code executions. For instance you may randomly dump some classes from the client on a random interval, As for the server you aay use a HashMap<String, String> to store the classname->hash. JVM actually stripes some matter information from the class in the runtime and it's hard to emulate such effect. This wont take much bandwidth, since these "challenges" are sent randomly.
- Add some 0-knowledge checks like adding a "challenge" entity, and send the render stack frame to the server. The server compares the colleced frames against and ban the players with different frame. By 0-knowledge I mean a hack client is nearly impossible to obtain the correct frame without mixin the render pipeline, which mess it up.
